scripts/utils/printing_tools.js

```diff
@@ -0,0 +1,52 @@
+const BN = require("bn.js")
+const bnOne = new BN(1)
+const bnTwo = new BN(2)
+const bnTen = new BN(10)
+const bn256 = new BN(256)
+const bnMaxUint = bnTwo.pow(bn256).sub(bnOne)
+
+/**
+ * A generalized version of "toWei" for tokens with an arbitrary amount of digits.
+ * If the decimal representation has more digits than the maximum amount possible, then the extra digits are truncated.
+ * @param {string} amount User-friendly representation for the amount of some ERC20 token
+ * @param {integer} digits Maximum number of decimal digits of the token
+ * @return {string} decimal representation of the number of token units corresponding to the input amount
+ */
+const fromUserToMachineReadable = function (amount, digits) {
+  if (digits < 0 || digits > 256) throw Error("Invalid number of digits for ERC20 token: " + digits.toString())
+  const re = /^(\d+)(\.(\d+))?$/ // a sequence of at least one digit (0-9), followed by optionally a dot and another sequence of at least one digit
+  const match = re.exec(amount)
+  if (match == null) throw Error("Failed to parse decimal representation of " + amount)
+  const decimalString = (match[3] || "0").padEnd(digits, "0")
+  const integerPart = new BN(match[1])
+  const decimalPart = new BN(decimalString)
+  const representation = integerPart.mul(bnTen.pow(new BN(digits))).add(decimalPart)
```

Comment:
Why is only the integer part shifted by a power of ten? Shouldn't it be the whole number?

Reply:
This converts user input to token internal representation. So, if the token has 3 decimals, this parses:
"12.01" -> "12010"
This script works in this way:
1. split number into integer part ("12") and decimal part ("01")
2. pad decimal part with zeroes to the right to get as many digits as there are decimals ("01" -> "010")
3. convert both to BN ("12", "010" -> 12, 10)
4. the integer part is "worth" 10^digits token units, while thanks to the padding the decimal part already represents units (so total amount of unit in the example is 12*10^3 + 10 = 12010)
5. convert to string ("12010")

All the rest of the code is just error checks and the special case of a token with zero decimals.

Reply:
okay thanks... I missed the smart padding.